core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -823,35 +833,138 @@ protected void validateAddedDVs(
+    DeleteFileSet conflictingDVs = dvsWithReferencedDataFile(spec, dataFile);
+    PositionDeleteIndex mergedPositions =
+        mergeConflictingDVs(Iterables.concat(conflictingDVs, committedDVsForDataFile.getValue()));
+    try {
+      DVFileWriter dvFileWriter =
+          new BaseDVFileWriter(
+              OutputFileFactory.builderFor(ops(), spec(specId), FileFormat.PUFFIN, 1, 1).build(),
+              path -> null);
+      dvFileWriter.delete(dataFile, mergedPositions, spec, committedDV.partition());
+      dvFileWriter.close();
+      DeleteWriteResult result = dvFileWriter.result();
+      DeleteFile mergedDV = Iterables.getOnlyElement(result.deleteFiles());
+      return new MergedDVResult(
+          new PendingDeleteFile(mergedDV), committedDVsForDataFile.getValue(), conflictingDVs);
     } catch (IOException e) {
-      throw new UncheckedIOException(e);
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  private static class MergedDVResult {
+    private final DeleteFile mergedDV;
+    private final Set<DeleteFile> committedDVs;
+    private final Set<DeleteFile> conflictingDVs;
+
+    MergedDVResult(
+        DeleteFile mergedDV, Set<DeleteFile> committedDVs, Set<DeleteFile> conflictingDVs) {
+      this.mergedDV = mergedDV;
+      this.committedDVs = committedDVs;
+      this.conflictingDVs = conflictingDVs;
+    }
+
+    public DeleteFile mergedDV() {
+      return mergedDV;
     }
+
+    public Set<DeleteFile> committedDVs() {
+      return committedDVs;
+    }
+
+    public Set<DeleteFile> conflictingDVs() {
+      return conflictingDVs;
+    }
+  }
+
+  private PositionDeleteIndex mergeConflictingDVs(Iterable<DeleteFile> conflictingDVs) {
+    Iterator<DeleteFile> confictingDVIterator = conflictingDVs.iterator();
+    PositionDeleteIndex mergedPositions = readDV(confictingDVIterator.next());
+    confictingDVIterator.forEachRemaining(dv -> mergedPositions.merge(readDV(dv)));
+    return mergedPositions;
+  }
+
+  private void replaceConflictingWithMergedDV(MergedDVResult mergedDVResult) {
+    DeleteFile mergedDV = mergedDVResult.mergedDV();
+    Set<DeleteFile> conflictingDVs = mergedDVResult.conflictingDVs();
+    // Remove the committed DVs from metadata
+    mergedDVResult.committedDVs().forEach(this::delete);
+    DeleteFileSet deleteFilesForSpec = newDeleteFilesBySpec.get(mergedDV.specId());
+    // Remove the pending conflicting DVs
+    deleteFilesForSpec.removeAll(conflictingDVs);
+    // Add the merged DV
+    deleteFilesForSpec.add(mergedDV);
+  }
+
+  private PositionDeleteIndex readDV(DeleteFile dv) {
+    LOG.trace("Opening DV file {}", dv.location());
+    InputFile inputFile = EncryptingFileIO.combine(ops().io(), ops().encryption()).newInputFile(dv);
+    long offset = dv.contentOffset();
+    int length = dv.contentSizeInBytes().intValue();
+    byte[] bytes = IOUtil.readBytes(inputFile, offset, length);
+    return PositionDeleteIndex.deserialize(bytes, dv);
+  }
```

Comment:
Couldn't use DeleteLoader since it's in iceberg-data, I think it maybe worth exposing this API on the Deletes class?